MediaBrowser.Controller/SyncPlay/Queue/PlayQueueManager.cs

```diff
@@ -0,0 +1,596 @@
+                    // Restoring playing item.
+                    SetPlayingItemByPlaylistId(playingItem.PlaylistItemId);
+                    return false;
+                }
+            }
+            else
+            {
+                return false;
+            }
+        }
+
+        /// <summary>
+        /// Moves an item in the playlist to another position.
+        /// </summary>
+        /// <param name="playlistItemId">The item to move.</param>
+        /// <param name="newIndex">The new position.</param>
+        /// <returns><c>true</c> if the item has been moved; <c>false</c> otherwise.</returns>
+        public bool MovePlaylistItem(string playlistItemId, int newIndex)
+        {
+            var list = GetPlaylistAsList();
+            var playingItem = list[PlayingItemIndex];
+
+            var playlistIds = list.Select(queueItem => queueItem.PlaylistItemId).ToList();
+            var oldIndex = playlistIds.IndexOf(playlistItemId);
+            if (oldIndex < 0) {
+                return false;
+            }
+
+            var queueItem = list[oldIndex];
+            list.RemoveAt(oldIndex);
+            newIndex = newIndex > list.Count() ? list.Count() : newIndex;
+            newIndex = newIndex < 0 ? 0 : newIndex;
+            list.Insert(newIndex, queueItem);
+
+            LastChange = DateTime.UtcNow;
+            PlayingItemIndex = list.IndexOf(playingItem);
+            return true;
+        }
+
+        /// <summary>
+        /// Resets the playlist to its initial state.
+        /// </summary>
+        public void Reset()
+        {
+            ProgressiveId = 0;
+            SortedPlaylist.Clear();
+            ShuffledPlaylist.Clear();
+            PlayingItemIndex = -1;
+            ShuffleMode = GroupShuffleMode.Sorted;
+            RepeatMode = GroupRepeatMode.RepeatNone;
+            LastChange = DateTime.UtcNow;
+        }
+
+        /// <summary>
+        /// Sets the repeat mode.
+        /// </summary>
+        /// <param name="mode">The new mode.</param>
+        public void SetRepeatMode(string mode)
+        {
+            switch (mode)
+            {
+                case "RepeatOne":
+                    RepeatMode = GroupRepeatMode.RepeatOne;
+                    break;
+                case "RepeatAll":
+                    RepeatMode = GroupRepeatMode.RepeatAll;
+                    break;
+                default:
+                    RepeatMode = GroupRepeatMode.RepeatNone;
+                    break;
+            }
+
+            LastChange = DateTime.UtcNow;
+        }
+
+        /// <summary>
+        /// Sets the shuffle mode.
+        /// </summary>
+        /// <param name="mode">The new mode.</param>
+        public void SetShuffleMode(string mode)
+        {
+            switch (mode)
+            {
+                case "Shuffle":
+                    ShufflePlaylist();
+                    break;
+                default:
+                    SortShuffledPlaylist();
```

Comment:
Whereas if i shuffled again, I would expect it to re-shuffle, not shuffling doesn't have to be repeated.
```suggestion
                if (!ShuffleMode.Equals(mode))
                { 
                    SortShuffledPlaylist();
                }
```

Reply:
The new changes should allow for re-shuffling when requested now.